cloudbridge/cloud/providers/aws/helpers.py

```diff
@@ -0,0 +1,260 @@
+        See issue: https://github.com/boto/boto3/issues/1268
+        '''
+        client = self.boto_conn.meta.client
+        list_op = self._get_list_operation()
+        if client.can_paginate(list_op):
+            log.debug("Supports server side pagination. Server will"
+                      " limit and page results.")
+            return self._get_paginated_results(limit, marker, collection)
+        else:
+            log.debug("Does not support server side pagination. Client will"
+                      " limit and page results.")
+            # Do not limit, let the ClientPagedResultList enforce limit
+            return (None, collection)
+
+    def list(self, limit=None, marker=None, collection=None):
+        collection = collection or self.boto_collection.filter()
+        resume_token, boto_objs = self._make_query(collection, limit, marker)
+
+        # Wrap in CB objects.
+        results = [self.cb_resource(self.provider, obj) for obj in boto_objs]
+
+        if resume_token:
+            log.debug("Received a resume token, using server pagination.")
+            return ServerPagedResultList(is_truncated=True,
+                                         marker=resume_token,
+                                         supports_total=False,
+                                         data=results)
+        else:
+            log.debug("Did not received a resume token, will page in client"
+                      " if necessary.")
+            return ClientPagedResultList(self.provider, results,
+                                         limit=limit, marker=marker)
+
+    def find(self, filter_name, filter_value, limit=None, marker=None):
+        '''
+        Returns a list of resources by filter
+
+        :param str filter_name: Name of the filter to use
+        :param str filter_value: Value to filter with
+        '''
+        collection = self.boto_collection
+        collection = collection.filter(Filters=[{
+            'Name': filter_name,
+            'Values': [filter_value]
+            }])
+        return self.list(limit=limit, marker=marker, collection=collection)
+
+    def create(self, boto_method, **kwargs):
+        '''
+        Creates a resource
+
+        :param str boto_method: AWS Service method to invoke
+        :param object kwargs: Arguments to be passed as-is to
+            the service method
+        '''
+        trimmed_args = trim_empty_params(kwargs)
+        result = getattr(self.boto_conn, boto_method)(**trimmed_args)
+        if isinstance(result, list):
+            return [self.cb_resource(self.provider, obj)
+                    for obj in result if obj]
+        else:
+            return self.cb_resource(self.provider, result) if result else None
+
+    def delete(self, resource_id):
+        '''
+        Deletes a resource by id
+
+        :param str id: ID of the resource
+        '''
+        res = self.get(resource_id)
+        if res:
+            res.delete()
```

Comment:
I prefer removing the bool for two reasons: One is that it seems redundant to return a bool if delete didn't throw an exception, which it should if the delete didn't succeed I think. The second is that only AWS returns a bool from delete.

Reply:
boto3 is inconsistent in what it returns from a delete action. For example: 
- Security Groups delete doesn't return anything: http://boto3.readthedocs.io/en/stable/reference/services/ec2.html#EC2.SecurityGroup.delete
- Instance terminate returns a pretty comprehensive dict: http://boto3.readthedocs.io/en/stable/reference/services/ec2.html#EC2.Instance.terminate
- VPC delete doesn't return anything: http://boto3.readthedocs.io/en/stable/reference/services/ec2.html#EC2.Vpc.delete

Our OpenStack side of the implementation makes a best effort guess, for example: https://github.com/gvlproject/cloudbridge/blob/master/cloudbridge/cloud/providers/openstack/services.py#L876

Perhaps returning a `bool` is filled with peril in the long run because we're taking on perhaps more than necessary but it certainly is nice to get some degree to confirmation in the client from calling a `delete` method vs. having to do `delete` followed by `get` to check if the action succeeded.

Reply:
> Perhaps returning a bool is filled with peril in the long run because we're taking on perhaps more than necessary but it certainly is nice to get some degree to confirmation in the client from calling a delete method vs. having to do delete followed by get to check if the action succeeded.

I think the problem here is that whether confirmation is needed or not is situation specific. There may be cases where, if the provider can't honour the request to delete the object, the user can't do much about it anyway, so they'll just have to continue. Incurring an unnecessary performance hit for a forced check seems spurious in this case, and possibly undesirable.

A good example of this is instance.terminate(), soon to be renamed instance.delete(). We expect users to use `wait_for(InstanceState.DELETED)` in this case anyway. Most often, they may just issue instance.delete(), go do something else, then check back later whether the instance was deleted. So in this case, forcing a wait would be very problematic.

Perhaps we can accept a parameter called, wait=True or verify=True or something, so that the method will block till the delete is complete, or throw an exception. That is, internally do a `wait_for([ObjState.DELETED, ObjState.UNKNOWN]).`